File: 프로그래머스/2/42586. 기능개발/기능개발.cpp
#include <string>
#include <cmath>
#include <vector>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    
    vector<int> answer;
    vector<int> left_days;
    
    // 남은 일수를 계산하여 left_days에 추가
    for (size_t i = 0; i < progresses.size(); ++i) {
        left_days.push_back(ceil((100.0 - progresses[i]) / speeds[i]));
    }
    
    int max_day = left_days[0];
    int cnt = 1;

    // left_days를 순회하며 배포 일자를 정리
    for (size_t i = 1; i < left_days.size(); ++i) {
        if (left_days[i] > max_day) {
            answer.push_back(cnt);
            cnt = 1;
            max_day = left_days[i];
        } else {
            ++cnt;
        }
    }
    
    // 마지막 배포 추가
    answer.push_back(cnt);
    return answer;
}

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.ArrayList;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        
        int n = progresses.length;
        int[] remainingDays = new int[n];

        for (int i = 0; i < n; i++) {
            remainingDays[i] = (int) Math.ceil((100.0 - progresses[i]) / speeds[i]);
        }


        ArrayList<Integer> result = new ArrayList<>();
        
        int currentMaxDay = remainingDays[0];
        int count = 1;

        for (int i = 1; i < n; i++) {
            if (remainingDays[i] <= currentMaxDay) {
                count++;
            } else {
                result.add(count);
                count = 1;
                currentMaxDay = remainingDays[i];
            }
        }

        result.add(count);

        return result.stream().mapToInt(Integer::intValue).toArray();
    }
}
